server/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java

```diff
@@ -156,34 +166,97 @@ public boolean addOrNotify(Translog.Location location, Consumer<Boolean> listene
         return true;
     }
 
+    /**
+     * Add a listener for refreshes, calling it immediately if the location is already visible. If this runs out of listener slots then it
+     * forces a refresh and calls the listener immediately as well. The checkpoint cannot be greater than the processed local checkpoint.
+     *
+     * @param checkpoint the seqNo checkpoint to listen for
+     * @param listener for the refresh.
+     * @return did we call the listener (true) or register the listener to call later (false)?
+     */
+    public boolean addOrNotify(long checkpoint, ActionListener<Void> listener) {
+        assert checkpoint >= SequenceNumbers.NO_OPS_PERFORMED;
+        if (checkpoint <= lastRefreshedCheckpoint) {
+            listener.onResponse(null);
+            return true;
+        }
+        long maxIssuedSequenceNumber = maxIssuedSeqNoSupplier.getAsLong();
+        if (checkpoint > maxIssuedSequenceNumber) {
+            IllegalArgumentException e = new IllegalArgumentException("Cannot wait for unissued seqNo checkpoint [wait_for_checkpoint="
+                + checkpoint + ", max_issued_seqNo=" + maxIssuedSequenceNumber + "]");
+            listener.onFailure(e);
+            return true;
+        }
```

Comment:
It would be safer to check if the default refresh interval is set ? Or at least that it's not set to -1 ?

Reply:
I added validation in SearchService.